sdk/daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Reader.scala

```diff
@@ -36,50 +40,66 @@ object ArchivePayload {
+    if (schemaMode)
+      Right(())
+    else
+      com.daml.SafeProto.ensureNoUnknownFields(m).left.map(Error.Parsing)
 
-        } yield payload
+  /* Convert an Archive proto message into a scala ArchivePayload.
+   *
+   * Checks that the hash of the payload matches the hash field, that the version is
+   * supported, and if `schemaMode` is `false`, ensures the message has no unknown fields.
+   */
+  def readArchive(
+      lf: DamlLf.Archive,
+      schemaMode: Boolean = false,
+  ): Either[Error, ArchivePayload] = for {
+    _ <- validateUnknownFields(lf, schemaMode)
+    theirHash <- lf.getHashFunction match {
+      case DamlLf.HashFunction.SHA256 =>
+        PackageId
+          .fromString(lf.getHash)
+          .left
+          .map(err => Error.Parsing("Invalid hash: " + err))
       case DamlLf.HashFunction.UNRECOGNIZED =>
         Left(Error.Parsing("Unrecognized hash function"))
     }
-  }
+    ourHash = MessageDigestPrototype.Sha256.newDigest
+      .digest(lf.getPayload.toByteArray)
+      .map("%02x" format _)
+      .mkString
+    _ <- Either.cond(
+      theirHash == ourHash,
+      (),
+      Error.Parsing(s"Mismatching hashes! Expected $ourHash but got $theirHash"),
+    )
+    proto <- ArchivePayloadParser.fromByteString(lf.getPayload)
+    payload <- readArchivePayload(theirHash, proto, schemaMode)
 
-  // Validate hash and version of a DamlLf.ArchivePayload
-  @throws[Error.Parsing]
+  } yield payload
+
+  /* Converts a DamlLf.ArchivePayload Protocol Buffer message to a Scala ArchivePayload.
+   *
+   * Checks that the payload's version is supported and, if `schemaMode` is `false`,
+   * ensures the message has no unknown fields.
+   */
   def readArchivePayload(
       hash: PackageId,
       lf: DamlLf.ArchivePayload,
-  ): Either[Error, ArchivePayload] =
-    lf.getSumCase match {
-      case DamlLf.ArchivePayload.SumCase.DAML_LF_1 =>
-        Lf1PackageParser
-          .fromByteString(lf.getDamlLf1)
-          .map(
-            ArchivePayload.Lf1(hash, _, LanguageMinorVersion(lf.getMinor))
-          )
-      case DamlLf.ArchivePayload.SumCase.DAML_LF_2 =>
-        Right(ArchivePayload.Lf2(hash, lf.getDamlLf2, LanguageMinorVersion(lf.getMinor)))
-      case DamlLf.ArchivePayload.SumCase.SUM_NOT_SET =>
-        Left(Error.Parsing("Unrecognized or Unsupported LF version"))
-    }
+      schemaMode: Boolean,
+  ): Either[Error, ArchivePayload] = lf.getSumCase match {
+    case DamlLf.ArchivePayload.SumCase.DAML_LF_1 =>
+      for {
+        _ <- validateUnknownFields(lf, schemaMode)
+        pkg <- lf1PackageParser.fromByteString(lf.getDamlLf1)
+      } yield ArchivePayload.Lf1(hash, pkg, LanguageMinorVersion(lf.getMinor))
+    case DamlLf.ArchivePayload.SumCase.DAML_LF_2 =>
+      for {
+        _ <- validateUnknownFields(lf, schemaMode)
+        minor = LanguageMinorVersion(lf.getMinor)
+      } yield ArchivePayload.Lf2(hash, lf.getDamlLf2, minor)
```

Comment:
This deviates from the 3.4 and main changes - on those branches `lf.getDamlLf2` returns a `ByteString`, however on the 3.3 branch we return a `DamlLf2.Package`. So please review these changes with care.

Reply:
This mismatch comes from the different recursion limit we use for 2.1 and 2.2, since this function does not care of the minor version, we change for Archive to byteString to delay proto parsing.

Since we handle only 2.1 it should be fine.